esp-metadata/src/lib.rs

```diff
@@ -1,250 +1,48 @@
-    pub fn name(&self) -> String {
-        self.device.name.clone()
-    }
-
-    /// The CPU architecture of the device.
-    pub fn arch(&self) -> Arch {
-        self.device.arch
-    }
-
-    /// The core count of the device.
-    pub fn cores(&self) -> Cores {
-        self.device.cores
-    }
-
-    /// The peripherals of the device.
-    pub fn peripherals(&self) -> &[String] {
-        &self.device.peripherals
-    }
-
-    /// User-defined symbols for the device.
-    pub fn symbols(&self) -> &[String] {
-        &self.device.symbols
-    }
-
-    /// All configuration values for the device.
-    pub fn all(&self) -> impl Iterator<Item = &str> + '_ {
-        [
-            self.device.name.as_str(),
-            self.device.arch.as_ref(),
-            self.device.cores.as_ref(),
-        ]
-        .into_iter()
-        .chain(self.device.peripherals.iter().map(|s| s.as_str()))
-        .chain(self.device.symbols.iter().map(|s| s.as_str()))
-    }
-
-    /// Does the configuration contain `item`?
-    pub fn contains(&self, item: &str) -> bool {
-        self.all().any(|i| i == item)
-    }
-
-    /// Define all symbols for a given configuration.
-    pub fn define_symbols(&self) {
-        define_all_possible_symbols();
-        // Define all necessary configuration symbols for the configured device:
-        for symbol in self.all() {
-            println!("cargo:rustc-cfg={symbol}");
+/// Macro to get the start of the given memory region.
+///
+/// ```rust,no-run
+/// const DRAM_START: usize = esp_metadata::memory_region_start("DRAM");
+/// ```
+#[macro_export]
+macro_rules! memory_region_start {
+    ( $var:expr ) => {
+        const {
+            match usize::from_str_radix(env!(concat!("ESP_METADATA_REGION_", $var, "_START")), 10) {
+                Ok(val) => val,
+                Err(_) => {
+                    core::assert!(false, "Unable to parse memory region.");
+                    0
+                }
+            }
         }
-    }
+    };
 }
 
-/// Defines all possible symbols that _could_ be output from this crate
-/// regardless of the chosen configuration.
+/// Macro to get the end of the given memory region.
 ///
-/// This is required to avoid triggering the unexpected-cfgs lint.
-fn define_all_possible_symbols() {
-    // Used by our documentation builds to prevent the huge red warning banner.
-    println!("cargo:rustc-check-cfg=cfg(not_really_docsrs)");
-
-    for chip in Chip::iter() {
-        let config = Config::for_chip(&chip);
-        for symbol in config.all() {
-            // https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-check-cfg
-            println!("cargo:rustc-check-cfg=cfg({})", symbol);
+/// ```rust,no-run
+/// const DRAM_END: usize = esp_metadata::memory_region_end("DRAM");
+/// ```
+#[macro_export]
+macro_rules! memory_region_end {
+    ( $var:expr ) => {
+        const {
+            match usize::from_str_radix(env!(concat!("ESP_METADATA_REGION_", $var, "_END")), 10) {
```

Comment:
Could you also update the esp-config macros whilst you're here? AKA resolve: https://github.com/esp-rs/esp-hal/blob/4dcb1c71935f9ff6dafa4c26af7f9df29628d71d/esp-config/src/lib.rs#L25